Imagery view for time strip

shefalijoshi opened the pull request:
Resolves #2977
Work Remaining:
- [x] UI fine tuning by @charlesh88 
- [ ] Tests @shefalijoshi 
- [ ] Testing instructions @shefalijoshi 
- [ ] Hide image placeholder when actual image thumbnail is being displayed @shefalijoshi 
- [x] Ensure that images on the right side of the strip are displayed, even if they are clipped on the right side of the view @shefalijoshi 
- [ ] Handle RT Time Conductor changes 
![Screen Shot 2021-08-19 at 4 52 28 PM](https://user-images.githubusercontent.com/1056412/130158452-b47f5154-b6d3-4703-8e4d-5582d59b5cce.png)


### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?

### Author Checklist

* [x] Changes address original issue?
* [ ] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [ ] Testing instructions included in associated issue?

akhenry commented on src/plugins/imagery/components/ImageryViewLayout.vue:
```diff
@@ -478,6 +527,9 @@ export default {
 
             if (viewLargeAction && viewLargeAction.appliesTo(this.objectPath, this.currentView)) {
```
I realize you didn't write this, but I'm just wondering why we're checking `appliesTo` here. It's guaranteed to apply to the given `objectPath` and `currentView` because they were passed into `getActionsCollection`, which uses `appliesTo` to populate the collection from all available actions. I think we can get rid of this check.

akhenry commented on src/plugins/imagery/ImageryView.js:
```diff
@@ -11,6 +11,8 @@ export default class ImageryView {
     }
 
     show(element) {
+        let isCompact = this.objectPath.find(object => object.type === 'time-strip');
```
It really feels to me as though the timestrip view of imagery should be a completely separate view. It's really a completely different way of viewing images.

I think we should define a new view provider, with a `canView` function that basically includes the logic above. The `view` function of the new provider would return the much stripped-down view of images appropriate for the Time View. 

The `canView` function of the ImageryViewProvider would probably also need to be modified to make sure it does not apply when the view is embedded in a timestrip.

akhenry replied:
You can use the `PreviewAction` to load the regular image view when an image is clicked on (see comments on this below), and use the Selection API for communicating which image has been clicked to the regular image view in its preview window.

shefalijoshi replied:
Ah!! I had this as a separate view before, but given the viewLarge functionality I brought the view into the Imagery Layout.
If PreviewAction works similarly to viewLarge then it definitely should be a separate view. I'll make the change.

shefalijoshi replied:
Done.

akhenry commented on src/plugins/imagery/components/ImageryTimeView.vue:
```diff
@@ -0,0 +1,432 @@
+    },
+    methods: {
+        expand(index) {
+            this.$emit('expand', index);
```
This could use the `PreviewAction` which will load a given object's view from scratch (unlike the `viewLargeAction` which will expand an already rendered view).

shefalijoshi replied:
Good idea! I used this action.

akhenry reviewed (changes requested):
Wow, this is super cool. I played around with it and it works really well. I did notice that the view large behavior does not work in realtime mode. I click on the images and nothing happens (also causes some flickering in Chrome).

I have left some suggestions for how I think we can leverage the API to do some of the heavy lifting for us, and achieve a cleaner separation of the two different representations of imagery.

shefalijoshi commented:
@akhenry 
I've separated the Imagery View from the Imagery Timestrip View as follows:
1. Created a new Imagery Timestrip view provider that only works if 'time-strip' object is in the objectPath
2. Updated the original imagery view provider to only allow the view if the 'time-strip' object is *NOT* in the objectPath
3. Created a mixin and moved common code (imagery telemetry subscription/request etc) into it.
4. Added the mixin to the ImageryLayout view and the ImageryTimestrip views.
5. ImageryTimestrip view uses Preview action to view the images in large view.
6. Imagery Layout still uses viewLarge to show an expanded view

akhenry reviewed (changes requested):
Love it, really nice clean separation now. Just a couple of suggestions that I don't think should be too much work. Thanks for all the work here, it's looking really great.

shefalijoshi commented on src/plugins/imagery/pluginSpec.js:
```diff
@@ -116,51 +117,51 @@ describe("The Imagery View Layout", () => {
-                        }
-                    }
+                    "source": "url"
+                    // "relatedTelemetry": {
```
Yes, for now. There are no tests for relatedTelemetry and it was causing some tests to fail. I'll work with Jamie to add tests for Related Telemetry later.

shefalijoshi commented on src/plugins/imagery/ImageryView.js:
```diff
@@ -23,6 +23,7 @@ export default class ImageryView {
                 currentView: this
             },
             template: '<imagery-view-layout ref="ImageryLayout"></imagery-view-layout>'
```
Done

shefalijoshi commented on src/plugins/imagery/ImageryViewProvider.js:
```diff
@@ -37,8 +37,10 @@ export default function ImageryViewProvider(openmct) {
-        canView: function (domainObject) {
-            return hasImageTelemetry(domainObject);
+        canView: function (domainObject, objectPath) {
+            let isCompact = objectPath.find(object => object.type === 'time-strip');
```
Done

shefalijoshi commented on src/plugins/imagery/components/ImageryTimeView.vue:
```diff
@@ -0,0 +1,468 @@
+            delete this.unsubscribe;
+        }
+
+        clearInterval(this.resizeTimer);
```
Done.

akhenry replied:
`_.debounce` should be fine, so long as the user experience doesn't suck. It just means it won't resize at all until the user finishes resizing the outer component, but I think that's probably ok.

shefalijoshi commented on src/plugins/imagery/ImageryTimestripViewProvider.js:
```diff
@@ -0,0 +1,73 @@
+        name: 'Imagery Timestrip View',
+        cssClass: 'icon-image',
+        canView: function (domainObject, objectPath) {
+            let isCompact = objectPath.find(object => object.type === 'time-strip');
```
Yes, it was a bug. I've added the navigatedObject condition. It was being used elsewhere so I moved the logic into ApplicationRouter

shefalijoshi replied:
Done

akhenry commented on src/plugins/imagery/Imagery_View.js:
```diff
@@ -1,8 +1,8 @@
-import ImageryViewLayout from './components/ImageryViewLayout.vue';
+import ImageryView from './components/ImageryView.vue';
```
This could be imported as `ImageryViewComponent` to avoid the naming conflict

shefalijoshi replied:
Done

akhenry commented on src/plugins/imagery/Imagery_View.js:
```diff
@@ -1,8 +1,8 @@
 import Vue from 'vue';
 
-export default class ImageryView {
+export default class Imagery_View {
```
Let's just reference the ImageryView component as ImageryViewComponent in this file so we can remain consistent with our code standards.

shefalijoshi replied:
Done

akhenry commented on src/plugins/imagery/components/ImageryTimeView.vue:
```diff
@@ -91,7 +96,10 @@ export default {
 
-        clearInterval(this.resizeTimer);
+        if (this.imageryStripResizeObserver) {
+            this.imageryStripResizeObserver.disconnect();
```
👍

akhenry reviewed (changes requested):
Thank you for making all of these changes, it's looking great. Just one tiny naming thing then we're good to go.

akhenry commented on src/ui/router/ApplicationRouter.js:
```diff
@@ -136,6 +136,13 @@ class ApplicationRouter extends EventEmitter {
         this.handleLocationChange(hash.substring(1));
     }
 
+    isNavigatedObject(objectPath) {
```
This is in another PR which I am just about to merge - https://github.com/nasa/openmct/pull/4179/files#diff-0bff5a6f99db64e9b5322caa61d2aa0344eda15148a8fba7ed4bcc992d5254f6R39

shefalijoshi replied:
We shall see which one wins ;)

charlesh88 reviewed (changes requested):
Still need changes to the Time Conductor when in RT mode, see todos.

akhenry commented:
@shefalijoshi  Is this still outstanding? - 
> Handle RT Time Conductor changes

akhenry commented:
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? N - followup has been filed
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Has associated issue been labelled `unverified`? (only applicable if this PR closes the issue) Y

akhenry commented:
File followup issue to address outstanding work - https://github.com/nasa/openmct/issues/4227
